modules/frontend/tagsharding.go

```diff
@@ -0,0 +1,431 @@
+	tagSearchReq            interface {
+		start() uint32
+		end() uint32
+		newWithRange(start, end uint32) tagSearchReq
+		buildSearchTagRequest(subR *http.Request) (*http.Request, error)
+		buildTagSearchBlockRequest(*http.Request, string, int, int, *backend.BlockMeta) (*http.Request, error)
+	}
+)
+
+type tagResultCollector struct {
+	delegate   tagResultsHandler
+	mtx        sync.Mutex
+	err        error
+	statusCode int
+	statusMsg  string
+	ctx        context.Context
+}
+
+type searchTagSharder struct {
+	next                   http.RoundTripper
+	reader                 tempodb.Reader
+	overrides              overrides.Interface
+	tagShardHandlerFactory tagResultHandlerFactory
+	cfg                    SearchSharderConfig
+	logger                 log.Logger
+	parseRequest           parseRequestFunction
+}
+
+type tagResults struct {
+	response    string
+	statusCode  int
+	statusMsg   string
+	err         error
+	marshallErr error
+}
+
+func (r *tagResultCollector) shouldQuit() bool {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	if r.err != nil {
+		return true
+	}
+	if r.ctx.Err() != nil {
+		return true
+	}
+	if r.statusCode/100 != 2 {
+		return true
+	}
+	if r.delegate.shouldQuit() {
+		return true
+	}
+	return false
+}
+
+func (r *tagResultCollector) setStatus(statusCode int, statusMsg string) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	r.statusCode = statusCode
+	r.statusMsg = statusMsg
+}
+
+func (r *tagResultCollector) setError(err error) {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	r.err = err
+}
+
+func (r *tagResultCollector) addResponseToResult(ic io.ReadCloser) error {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+	return r.delegate.addResponse(ic)
+}
+
+func (r *tagResultCollector) Result() *tagResults {
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	response, err := r.delegate.marshalResult()
+	return &tagResults{
+		statusCode:  r.statusCode,
+		statusMsg:   r.statusMsg,
+		err:         r.err,
+		response:    response,
+		marshallErr: err,
+	}
+}
+
+func newTagResultCollector(ctx context.Context, factory tagResultHandlerFactory, limit int) *tagResultCollector {
+	return &tagResultCollector{
+		statusCode: http.StatusOK,
+		ctx:        ctx,
+		delegate:   factory(limit),
+	}
+}
+
+func (s searchTagSharder) httpErrorResponse(err error) *http.Response {
+	return &http.Response{
+		StatusCode: http.StatusBadRequest,
+		Body:       io.NopCloser(strings.NewReader(err.Error())),
+	}
+}
+
+// RoundTrip implements http.RoundTripper
+// execute up to concurrentRequests simultaneously where each request scans ~targetMBsPerRequest
+// until limit results are found
+func (s searchTagSharder) RoundTrip(r *http.Request) (*http.Response, error) {
+	requestCtx := r.Context()
+
+	tenantID, err := user.ExtractOrgID(requestCtx)
+	if err != nil {
+		return s.httpErrorResponse(err), nil
+	}
+
+	handler := newTagResultCollector(requestCtx, s.tagShardHandlerFactory, s.overrides.MaxBytesPerTagValuesQuery(tenantID))
```

Comment:
`MaxBytesPerTagValuesQuery` is defined only for `/tag/<tag>/values`, and not for `/tags`. This is changing the behaviour.

I'm not opposed to the change, but it should at least get documented.

Reply:
`MaxBytesPerTagValuesQuery` is also implemented in the ingester—this was fine since only a single ingester request was created. It'd be good to coalesce this limit in a single place.

Reply:
I put a TODO here, may be we can address it in a follow up PR?